### ISSUE TITLE
[Question] Prepared Statements and Multiple Rows with One Query

### ISSUE THREAD

**Issue by hellopat** [sidorares/node-mysql2#747]
Is there a way to insert multiple rows via a single prepared statement?

Something like the following:

`const results = await db.execute('INSERT INTO kv (key, value) VALUES ?', [["key", "value"], ["key2", "value2"]]);`

**Comment by sidorares** [comment]
not really, there must by exactly same amount of placeholders as number of input parameters you pass

Maybe if you pass json object and do some json manipulation on server side, but I never used JSON functions in mysql so not sure if it's easy to do json object -> multiple rows via single query (or call to stored procedure) - https://dev.mysql.com/doc/refman/5.7/en/json.html

**Comment by hellopat** [comment]
Not 100% sure I understand, but I was able to get it to work like this:

```
const results = await db.query(
  'INSERT INTO kv (k, v) VALUES ?',
  [[["key", "value"], ["key2", "value2"]]]
);
```

Could I possible make this a prepared statement by doing something like:

```
const params = [
  ["key", "value"],
  ["key2", "value2"]
];

const inList = params.map(row => '(?, ?)').join(', ');

const results = await db.excute(
  'INSERT INTO kv (k, v) VALUES (' + inList + ')',
  [params]
);
```

It doesn't work, but I may be missing something.

**Comment by sidorares** [comment]
this question is basically a duplicate of #742 See my comment -  https://github.com/sidorares/node-mysql2/issues/742#issuecomment-369807007

TLDR: `.execute()` and `.query()` are different even if api looks the same

**Comment by sidorares** [comment]
for your example you need to flatten pams list
`console.log('INSERT INTO kv (k, v) VALUES (' + inList + ')');` prints `INSERT INTO kv (k, v) VALUES ((?, ?), (?, ?))` so you need to call later it as 

```js
const results = await db.excute(
  'INSERT INTO kv (k, v) VALUES (' + inList + ')',
  [params].reduce((res, item) => res.concat(item))
);
``` 
because you want to pass 4 values - `[ 'key', 'value', 'key2', 'value2' ]` (otherwise you pass 2 arrays and they converted `toString()` and result is 'key,value' for first and 'key2,value2' for second

There is not much benefit in using prepared statements with this approach, you build query manually and kind of lose the separation of query and data prepared statements are designed to give you
